openghg/retrieve/_access.py

```diff
@@ -509,6 +510,69 @@ def get_obs_column(
         **kwargs,
     )
 
+    if max_level > max(obs_data.data.lev.values) + 1:
+        print(
+            f"passed max level is above max level in data ({max(obs_data.data.lev.values)+1}). Defaulting to highest level"
+        )
+        max_level = max(obs_data.data.lev.values) + 1
+
+    ## processing taken from acrg/acrg/obs/read.py get_gosat()
+    lower_levels = list(range(0, max_level))
+
+    prior_factor = (
+        obs_data.data.pressure_weights[dict(lev=list(lower_levels))]
+        * (1.0 - obs_data.data.xch4_averaging_kernel[dict(lev=list(lower_levels))])
+        * obs_data.data.ch4_profile_apriori[dict(lev=list(lower_levels))]
+    ).sum(dim="lev")
+
+    upper_levels = list(range(max_level, len(obs_data.data.lev.values)))
+    prior_upper_level_factor = (
+        obs_data.data.pressure_weights[dict(lev=list(upper_levels))]
+        * obs_data.data.ch4_profile_apriori[dict(lev=list(upper_levels))]
+    ).sum(dim="lev")
+
+    obs_data.data["mf_prior_factor"] = prior_factor
+    obs_data.data["mf_prior_upper_level_factor"] = prior_upper_level_factor
+    obs_data.data["mf"] = (
+        obs_data.data.xch4 - obs_data.data.mf_prior_factor - obs_data.data.mf_prior_upper_level_factor
+    )
+    obs_data.data["mf_repeatability"] = obs_data.data.xch4_uncertainty
+
+    # rt17603: 06/04/2018 Added drop variables to ensure lev and id dimensions are also dropped, Causing problems in footprints_data_merge() function
+    drop_data_vars = [
+        "xch4",
+        "xch4_uncertainty",
+        "lon",
+        "lat",
+        "ch4_profile_apriori",
+        "xch4_averaging_kernel",
+        "pressure_levels",
+        "pressure_weights",
+        "exposure_id",
+    ]
+    drop_coords = ["lev", "id"]
+
+    for dv in drop_data_vars:
+        if dv in obs_data.data.data_vars:
+            obs_data.data = obs_data.data.drop_vars(dv)
+    for coord in drop_coords:
+        if coord in obs_data.data.coords:
+            obs_data.data = obs_data.data.drop_vars(coord)
+
+    obs_data.data = obs_data.data.sortby("time")
+
+    obs_data.data.attrs["max_level"] = max_level
+    if species.upper() == "CH4":
+        obs_data.data.mf.attrs["units"] = "1e-9"
```

Comment:
Ultimately we will want the "units" attribute included but we should probably find a way to extract this from the data rather than setting this. Probably worth commenting out for now (for CH4 and CO2) but could you open a quick issue for this?